<h1>Оглавление<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Исследуйте-таблицы-—-выведите-первые-строки" data-toc-modified-id="Исследуйте-таблицы-—-выведите-первые-строки-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Исследуйте таблицы — выведите первые строки</a></span><ul class="toc-item"><li><span><a href="#Таблица-books" data-toc-modified-id="Таблица-books-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Таблица books</a></span></li><li><span><a href="#Таблица-authors" data-toc-modified-id="Таблица-authors-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>Таблица authors</a></span></li><li><span><a href="#Таблица-publishers" data-toc-modified-id="Таблица-publishers-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Таблица publishers</a></span></li><li><span><a href="#Таблица-ratings" data-toc-modified-id="Таблица-ratings-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>Таблица ratings</a></span></li><li><span><a href="#Таблица-reviews" data-toc-modified-id="Таблица-reviews-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>Таблица reviews</a></span></li><li><span><a href="#Выводы-по-главе" data-toc-modified-id="Выводы-по-главе-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>Выводы по главе</a></span></li></ul></li><li><span><a href="#Решение-задач" data-toc-modified-id="Решение-задач-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Решение задач</a></span><ul class="toc-item"><li><span><a href="#Задача-№1" data-toc-modified-id="Задача-№1-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>Задача №1</a></span></li><li><span><a href="#Задача-№2" data-toc-modified-id="Задача-№2-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>Задача №2</a></span></li><li><span><a href="#Задача-№3" data-toc-modified-id="Задача-№3-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>Задача №3</a></span></li><li><span><a href="#Задача-№4" data-toc-modified-id="Задача-№4-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>Задача №4</a></span></li><li><span><a href="#Задача-№5" data-toc-modified-id="Задача-№5-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>Задача №5</a></span></li><li><span><a href="#Выводы-по-главе" data-toc-modified-id="Выводы-по-главе-2.6"><span class="toc-item-num">2.6&nbsp;&nbsp;</span>Выводы по главе</a></span></li></ul></li><li><span><a href="#Выводы-по-исследованию" data-toc-modified-id="Выводы-по-исследованию-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Выводы по исследованию</a></span><ul class="toc-item"><li><span><a href="#Рекомендации" data-toc-modified-id="Рекомендации-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>Рекомендации</a></span></li></ul></li></ul></div>

# SQL

**Описание проекта**

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.
Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

**Задания**
1. Посчитайте, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитайте количество обзоров и среднюю оценку;
3. Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
4. Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
5. Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Как выполнить задание?**
- Опишите цели исследования;
- Исследуйте таблицы — выведите первые строки;
- Сделайте по одному SQL-запросу для решения каждого задания;
- Выведите результат каждого запроса в тетрадке;
- Опишите выводы по каждой из решённых задач.

**Цели исследования:**
1. Сделать запросы к базе данных и получить ответы на поставленные вопросы.
2. Найти интересные особенности и дать рекомендации по продаже книг.

## Исследуйте таблицы — выведите первые строки

In [1]:
# импортируем библиотеки для исследования
import pandas as pd
import psycopg2
from sqlalchemy import create_engine

Установим соединение с базой данных.

In [2]:
# устанавливаем параметры для доступа к базе данных
db_config = {'user': 'praktikum_student', # имя пользователя
             'pwd': 'Sdf4$2;d-d30pp', # пароль
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432, # порт подключения
             'db': 'data-analyst-final-project-db'} # название базы данных

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                         db_config['pwd'],
                                                         db_config['host'],
                                                         db_config['port'],
                                                         db_config['db'])

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

Соединение с базой данных было установлено.

Теперь исследуем таблицы, путем вывода первых строк таблиц.

### Таблица books

In [3]:
# напишем запрос к таблице books
query_books = ''' SELECT *
                  FROM books
                  LIMIT 5
              ''' 

In [4]:
# выполним sql-запрос
books = pd.io.sql.read_sql(query_books, con = engine)

In [5]:
# просмотрим таблицу books
books.head()

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

### Таблица authors

In [6]:
# напишем запрос к таблице authors
query_authors = ''' SELECT *
                    FROM authors
                    LIMIT 5
                ''' 

In [7]:
# выполним sql-запрос
authors = pd.io.sql.read_sql(query_authors, con = engine)

In [8]:
# просмотрим таблицу authors
authors.head()

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

### Таблица publishers

In [9]:
# напишем запрос к таблице publishers
query_publishers = ''' SELECT *
                       FROM publishers
                       LIMIT 5
                   ''' 

In [10]:
# выполним sql-запрос
publishers = pd.io.sql.read_sql(query_publishers, con = engine)

In [11]:
# просмотрим таблицу publishers
publishers.head()

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

### Таблица ratings

In [12]:
# напишем запрос к таблице ratings
query_ratings = ''' SELECT *
                    FROM ratings
                    LIMIT 5
                ''' 

In [13]:
# выполним sql-запрос
ratings = pd.io.sql.read_sql(query_ratings, con = engine)

In [14]:
# просмотрим таблицу ratings
ratings.head()

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

### Таблица reviews

In [15]:
# напишем запрос к таблице reviews
query_reviews = ''' SELECT *
                    FROM reviews
                    LIMIT 5
                ''' 

In [16]:
# выполним sql-запрос
reviews = pd.io.sql.read_sql(query_reviews, con = engine)

In [17]:
# просмотрим таблицу reviews
reviews.head()

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Первые строки таблиц были просмотрены. Проблем в них не было обнаружено.

### Выводы по главе

**Выводы:**
    
1. Соединение с базой данных было установлено.

2. Первые строки таблиц были исследованы. Проблем не было обнаружено.

## Решение задач

### Задача №1

**Посчитайте, сколько книг вышло после 1 января 2000 года.**

In [18]:
# напишем запрос к задаче
# посчитаем количество книг после 01.01.2000
query_1 = ''' SELECT COUNT(book_id) as cnt_books_since_2000
              FROM books
              WHERE publication_date > '2000-01-01'
          ''' 

In [19]:
# выполним sql-запрос
books_2000 = pd.io.sql.read_sql(query_1, con = engine)

In [20]:
# просмотрим таблицу c результатом
books_2000.head()

cnt_books_since_2000
0                   819

После 1 января 2000 года вышло 819 книг.

### Задача №2

**Для каждой книги посчитайте количество обзоров и среднюю оценку.**

In [21]:
# напишем запрос к задаче
# посчитаем в 1 подзапросе для каждой книги количество обзоров
# посчитаем в 2 подзапросе для каждой книги среднюю оценку
query_2 = ''' SELECT rev.book_id,
                     rev.title,
                     cnt_reviews,
                     avg_rating
              FROM 
                  (SELECT b.book_id, 
                          b.title,
                          COUNT(r.review_id) AS cnt_reviews
                   FROM books AS b
                   LEFT JOIN reviews AS r ON b.book_id = r.book_id
                   GROUP BY b.book_id, b.title) AS rev       
              LEFT JOIN     
                  (SELECT book_id,
                          ROUND(AVG(rating), 2) AS avg_rating
                   FROM ratings
                   GROUP BY book_id) AS rat ON rev.book_id = rat.book_id
              ORDER BY cnt_reviews DESC
   
          '''

In [22]:
# выполним sql-запрос
books_reviews = pd.io.sql.read_sql(query_2, con = engine)

In [23]:
# просмотрим таблицу c результатом
books_reviews

book_id                                              title  cnt_reviews  \
0        948                            Twilight (Twilight  #1)            7   
1        695  The Curious Incident of the Dog in the Night-Time            6   
2        299  Harry Potter and the Chamber of Secrets (Harry...            6   
3        696             The Da Vinci Code (Robert Langdon  #2)            6   
4        734                                   The Glass Castle            6   
..       ...                                                ...          ...   
995      672   The Cat in the Hat and Other Dr. Seuss Favorites            0   
996      808                            The Natural Way to Draw            0   
997      221                          Essential Tales and Poems            0   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel            0   
999      387                               Leonardo's Notebooks            0   

     avg_rating  
0          3.66  
1          4.08  
2          4.29  
3          3.83  
4          4.21  
..          ...  
995        5.00  
996        3.00  
997        4.00  
998        3.67  
999        4.00  

[1000 rows x 4 columns]

В нашей базе данных не так много обзоров на книги. Но лидер это Сумерки. С 7 обзорами это самая популярная книга по количеству обзоров. Другие лидеры по количеству обзоров это Загадочное ночное убийство собаки, Гарри Поттер и тайная комната, Код да Винчи, Стеклянный замок.

При этом видно, что популярные по количеству обзоров книги могут иметь не самую высокую оценку. Заметно, что некоторые книги вообще не имеют обзоров.

### Задача №3

**Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.**

In [24]:
# напишем запрос к задаче
# посчитаем для издательсв количество выпущенных книг толще 50 страниц
# сделаем сортировку по убыванию количества книг и отберем только 1 строку
query_3 = ''' SELECT p.publisher,
                     COUNT(book_id) AS cnt_books
              FROM publishers AS p
              FULL OUTER JOIN books AS b ON p.publisher_id = b.publisher_id
              WHERE num_pages > 50
              GROUP BY p.publisher
              ORDER BY cnt_books DESC
              LIMIT 1
          ''' 

In [25]:
# выполним sql-запрос
top_publisher = pd.io.sql.read_sql(query_3, con = engine)

In [26]:
# просмотрим таблицу c результатом
top_publisher

publisher  cnt_books
0  Penguin Books         42

Издательство Penguin Books издало 42 книги, толще 50 страниц.

### Задача №4

**Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.**

In [27]:
# напишем запрос к задаче
# найдем книги с 50 и более оценками, затем найдем авторов с такими книгами 
# сделаем сортировку по убыванию средней оценки книг  книг и отберем только 1 строку
query_4 = ''' SELECT a.author, 
                     AVG(avg_rating_book) AS avg_rating
              FROM authors AS a
              FULL OUTER JOIN books AS b ON a.author_id = b.author_id
              JOIN (SELECT b.book_id,
                                      ROUND(AVG(r.rating), 2) AS avg_rating_book,
                                      COUNT(r.rating_id) AS cnt_rating
                               FROM books AS b
                               FULL OUTER JOIN ratings AS r ON b.book_id = r.book_id
                               GROUP BY b.book_id
                               HAVING COUNT(r.rating_id) >= 50
                               ORDER BY cnt_rating DESC) AS rat ON b.book_id = rat.book_id
              GROUP BY a.author
              ORDER BY avg_rating DESC
              LIMIT 1
          ''' 

In [28]:
# выполним sql-запрос
top_author = pd.io.sql.read_sql(query_4, con = engine)

In [29]:
# просмотрим таблицу c результатом
top_author

author  avg_rating
0  J.K. Rowling/Mary GrandPré       4.285

Дж.К. Роулинг - писатель с самой высокой средней оценкой книг (с учетом только книг, у которых от 50 оценок).

### Задача №5

**Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.**

In [30]:
# напишем запрос к задаче
# найдем в 1 подзапросе пользователей, которые поставили больше 50 оценок
# в 2 подзапросе посчитаем количество обзоров от пользователей
# затем получим среднее число обзоров пользователей

query_5 = ''' SELECT ROUND(AVG(cnt_reviews), 2) avg_cnt_reviews
              FROM 
                 (SELECT username,
                         COUNT(review_id) AS cnt_reviews
                  FROM reviews
                  WHERE username IN (SELECT username
                                     FROM ratings
                                     GROUP BY username
                                     HAVING COUNT(rating_id) > 50)
                  GROUP BY username) as users
          ''' 

In [31]:
# выполним sql-запрос
top_users = pd.io.sql.read_sql(query_5, con = engine)

In [32]:
# просмотрим таблицу c результатом
top_users

avg_cnt_reviews
0            24.33

В среднем, пользователи, которые поставили больше 50 оценок книгам, делают 24 обзора книг.

### Выводы по главе

**Выводы:**
    
1. После 1 января 2000 года вышло 819 книг.

2. В нашей базе данных не так много обзоров на книги. Но лидер это Сумерки. С 7 обзорами это самая популярная книга по количеству обзоров. Другие лидеры по количеству обзоров это Загадочное ночное убийство собаки, Гарри Поттер и тайная комната, Код да Винчи, Стеклянный замок. При этом видно, что популярные книги могуть иметь не самую высокую оценку. Заметно, что некоторые книги вообще не имеют обзоров.

3. Издательство Penguin Books издало 42 книги, толще 50 страниц.

4. Дж.К. Роулинг - писатель с самой высокой средней оценкой книг (с учетом только книг, у которых от 50 оценок).

5. В среднем, пользователи, которые поставили больше 50 оценок книгам, делают 24 обзора книг.

## Выводы по исследованию

Мы сделали запросы к базе данных и получили ответы на поставленные вопросы.

- Посчитайте, сколько книг вышло после 1 января 2000 года;

**Ответ: После 1 января 2000 года вышло 819 книг.**

- Для каждой книги посчитайте количество обзоров и среднюю оценку;

**Ответ: Мы посчитали количество обзоров и среднюю оценку для каждой книги. В нашей базе данных не так много обзоров на книги. Но лидер это Сумерки. С 7 обзорами это самая популярная книга по количеству обзоров. Другие лидеры по количеству обзоров это Загадочное ночное убийство собаки, Гарри Поттер и тайная комната, Код да Винчи, Стеклянный замок. При этом видно, что популярные по количеству обзоров книги могут иметь не самую высокую оценку. Заметно, что некоторые книги вообще не имеют обзоров.**

- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

**Ответ: Издательство Penguin Books издало 42 книги, толще 50 страниц.**

- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

**Ответ: Дж.К. Роулинг - писатель с самой высокой средней оценкой книг (с учетом только книг, у которых от 50 оценок).**

- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

**Ответ: В среднем, пользователи, которые поставили больше 50 оценок книгам, делают 24 обзора книг.**

### Рекомендации

1. Стоит сосредоточиться на продаже популярных по количеству обзоров книг. При этом популярные по количеству обзоров, книги могут иметь не самые высокие оценки пользователей.
2. Стоит обязательно продавать книги издательства Penguin Books. Это самое крупное издательство. Предполагаем, что достаточно много людей будет искать книги этого издательства.
3. Также стоит обязательно продавать книги автора Дж.К. Роулинг. Это писатель с самой высокой средней оценкой книг. Предполагаем, что читателит будут искать книги с высокой средней оценкой.
4. Стоит поощрять пользователей оставлять обзоры на книги. Это поможет нам выявлять популярные книги.
5. В нашем сервисе есть активные обзорщики книг, которые сделали от 24 обзоров книг. Стоит поощрять таких читателей, это также поможет нам выявлять популярные книги.